### PR TITLE
Windows: Graphdriver should reject create of layer w/o parent

### PR DESCRIPTION
**Pull request by swernli**
Signed-off-by: Stefan J. Wernli swernli@microsoft.com

@jhowardmsft @jstarks @stevvooe @icecrime @dmp42
Windows Graphdriver will allow creation of layers with missing parents.  This can lead to layers that fail to get cleaned up when performing a pull of a layer from a registry to a host missing the matching base image (such as a pull of a TP3 image onto a non-TP3 host).  This change allows the operation to fail earlier and get cleaned up properly.

### CODE REVIEW

File: daemon/graphdriver/windows/windows.go [hunk 1]
```diff
@@ -145,8 +145,16 @@ func (d *Driver) Create(id, parent string) error {
 			return err
 		}
 	} else if os.IsNotExist(err) {
-		// If the parent doesn't exist, this must be a special creation for an image
-		// registered at an alternate location. Use the parent id as the alternate ID.
+		if !d.Exists(parent) {
+			// The graphdriver did not find the parent in the central store.
+			if err2 := hcsshim.DestroyLayer(d.info, id); err2 != nil {
+				logrus.Warnf("Failed to DestroyLayer %s: %s", id, err)
```
Comment (calavera) [inline daemon/graphdriver/windows/windows.go]: did you mean to log `err2`?

File: daemon/graphdriver/windows/windows.go [hunk 2]
```diff
@@ -145,8 +145,16 @@ func (d *Driver) Create(id, parent string) error {
...
+			}
+			return fmt.Errorf("Cannot create layer with missing parent %s: %s", parent, err)
+		}
+		// If the parent doesn't exist but the graphdriver found it, this must be a
+		// special creation for an image registered at an alternate location.
+		// Use the parent id as the alternate ID.
 		if err := d.setId(id, parent); err != nil {
 			if err2 := hcsshim.DestroyLayer(d.info, id); err2 != nil {
 				logrus.Warnf("Failed to DestroyLayer %s: %s", id, err)
```
Comment (calavera) [inline daemon/graphdriver/windows/windows.go]: now that I notice it, this is also incorrect, isn't it?

File: daemon/graphdriver/windows/windows.go (hunk 1 above)
Comment (swernli) [inline daemon/graphdriver/windows/windows.go]: Yeah, good catch!

File: daemon/graphdriver/windows/windows.go (hunk 2 above)
Comment (swernli) [inline daemon/graphdriver/windows/windows.go]: I am an innocent victim of copy paste.

File: daemon/graphdriver/windows/windows.go [hunk 3]
```diff
@@ -140,22 +140,30 @@ func (d *Driver) Create(id, parent string) error {
...
+		// If the parent doesn't exist but the graphdriver found it, this must be a
+		// special creation for an image registered at an alternate location.
+		// Use the parent id as the alternate ID.
 		if err := d.setId(id, parent); err != nil {
 			if err2 := hcsshim.DestroyLayer(d.info, id); err2 != nil {
-				logrus.Warnf("Failed to DestroyLayer %s: %s", id, err)
+				logrus.Warnf("Failed to DestroyLayer %s: %s", id, err2)
 			}
 			return err
 		}
```
Comment (stevvooe) [inline daemon/graphdriver/windows/windows.go]: Should there be a call to `DestroyLayer` if there is an error but the other conditions are not met? If there is a `NotExist` error here, this method returns `nil`. Is that intended?

**Comment by stevvooe** [comment]
This code would be much clearer if the non-error flow was left out of the if statement:

```
 	if _, err := os.Lstat(d.dir(parent)); err != nil {
		if os.IsNotExist(err) {

			if !d.Exists(parent) {
				// The graphdriver did not find the parent in the central store.
				if err2 := hcsshim.DestroyLayer(d.info, id); err2 != nil {
					logrus.Warnf("Failed to DestroyLayer %s: %s", id, err2)
				}
				return fmt.Errorf("Cannot create layer with missing parent %s: %s", parent, err)
			}

			// If the parent doesn't exist but the graphdriver found it, this must be a
			// special creation for an image registered at an alternate location.
			// Use the parent id as the alternate ID.
	 		if err := d.setId(id, parent); err != nil {
	 			if err2 := hcsshim.DestroyLayer(d.info, id); err2 != nil {
					logrus.Warnf("Failed to DestroyLayer %s: %s", id, err2)
	 			}
	 			return err
	 		}
	 	} else {
 			if err2 := hcsshim.DestroyLayer(d.info, id); err2 != nil {
				logrus.Warnf("Failed to DestroyLayer %s: %s", id, err2)
 			}
 			return err
 		}
 	}

 	if err := d.setLayerChain(id, layerChain); err != nil {
		if err2 := hcsshim.DestroyLayer(d.info, id); err2 != nil {
			logrus.Warnf("Failed to DestroyLayer %s: %s", id, err2)
		}
		return err
	}
```

I'm not sure if I've converted this correctly, but written this way, the error handling is in the clearly separated from the standard flow. Put this way, it is much easier to see that we may have a condition where an error happens but is not reported.

File: daemon/graphdriver/windows/windows.go (hunk 3 above)
Comment (swernli) [inline daemon/graphdriver/windows/windows.go]: I'm not sure I understand what condition you are talking about.  If the error is a NotExist, then the code will check if the driver is aware of it.  If it also doesn't exist in the driver, we return an error, but if it does exist in the driver then we try to set the ID.  Which case is the one that is problematic?

**Comment by swernli** [comment]
I may be misunderstanding here; which error is not reported?  In both the original code and your refactored code, each return statement has an err variable that has been recently set.  Which line has the return that are worried will not have the err variable set?

File: daemon/graphdriver/windows/windows.go (hunk 3 above)
Comment (stevvooe) [inline daemon/graphdriver/windows/windows.go]: But, if neither of those conditions are true, `nil` is returned. Is that intended?

**Comment by stevvooe** [comment]
@swernli If `os.IsNotExist(err)` is true, but neither of `!d.Exists(parent)` or `d.setId(id, parent)` is true or errors out, an error occurred but is not returned. Is that the intended behavior?

**Comment by swernli** [comment]
Ah, I understand!  Yes, this is intended.  <code>os.IsNotExist(err) </code> being true is not an error on it's own, it simply indicates that we are in a special scenario, which may or may not be an error.

**Comment by stevvooe** [comment]
@swernli Great!

**Comment by swernli** [comment]
Actually, after looking through this code again, I realized that the special case I'm trying to preserve doesn't actually end up here any more.  With the graph refactoring that took place, this special case of needing to call <code>setId</code> is now handled by <code>RestoreCustomImages</code> down on line 442.  Given that, this error case can be significantly simplified.  I'm going to refactor it in light of that and resubmit.

**Comment by stevvooe** [comment]
LGTM

@swernli Much nicer!

**Comment by jfrazelle** [comment]
LGTM

**Pull request merged**